src/systems/bread/alpha/core.cljc

```diff
@@ -166,13 +166,50 @@
   (let [forms (map #(cons `add-hook %) forms)]
     `(-> ~app' ~@forms)))
 
+(defprotocol Effect
+  "Protocol for encapsulating side-effects"
+  (effect!
+    [this req]))
+
+(extend-protocol Effect
+  clojure.lang.Fn
+  (effect!
+    ([f req]
+     (f req)))
+
+  clojure.lang.PersistentVector
+  (effect!
+    ([v req]
+     (let [[f & args] v]
+       (apply f req args))))
+
+  java.util.concurrent.Future
+  (effect!
+    ([fut _]
+     (deref fut))))
+
 (defn add-effect
-  "Adds the (presumably effectful) function f as a callback to the special
-  :hook/effects hook. Accepts an optional options map."
-  ([app f options]
-   (add-hook app :hook/effects f options))
-  ([app f]
-   (add-hook app :hook/effects f {})))
+  "Adds e as an Effect to be run during the apply-effects lifecycle phase."
+  [req e]
+  (update req ::effects (comp vec conj) e))
+
+(defn effect?
+  "Whether x implements (satisfies) the Effect protocol"
+  [x]
+  (satisfies? Effect x))
+
+(defn- apply-effects [req]
+  (loop [{data ::data [effect & effects] ::effects :as req} req]
+    (if-not (effect? effect)
+      req
+      (let [;; DO THE THING!
+            {new-data ::data new-effects ::effects} (effect! effect req)
+            data (or new-data data)
+            effects (or new-effects effects)
+            req (assoc req ::data data ::effects effects)]
+        (if-not (seq effects)
+          req
+          (recur req))))))
```

Comment:
I actually tried something similar to each of these approaches! 😄

The `cond` one doesn't work because it loops infinitely. Notice the `:else` clause is actually subtly different logic from my implementation. The destructuring of `[effect & effects]` in the loop binding is important - your suggestion is close but it doesn't actually iterate "down" to zero Effects. Crucially, the `(or new-effects effects)` needs access to _both_ the decremented `effects` from the old vector and the new ones from calling `effect!` (if any).

Also, `merge-with` isn't quite what I want because it makes it too easy to return something like `{::config ...}` from an Effect and change things that Effects shouldn't. So that's why I did an explicit `assoc` with just those two keys. I supposed I could shave off a line or two with a clever application of `merge-with` and `select-keys` or something, but the logic here is subtle enough that I think this actually benefits from being a bit more verbose and explicit.

The `reduce` doesn't work because it doesn't take into account any new Effects that get returned from `effect!` - it binds to the vector that's there when it's first called and reduces over those Effects only. I had the same intuition - that it's a simple reduction over Effects - but it turns out that the ability to return new Effects makes it more complicated than that! The ability to enqueue new Effects is important to me, and worth a little complexity in the implementation.

I appreciate where this is coming from though, and I'm still open to the possibility that there's a better way to express it! 😄  But this slightly awkward loop is the clearest implementation I've come up with so far. Thanks so much for taking the time!!

Reply:
Drat! I knew it was too easy... of course, you would have discarded the other ideas :see_no_evil:  

Sorry about this digression I should have just cloned and tested. Thanks for the explanation!